doc/designs/quic-design/error-handling.md

```diff
@@ -0,0 +1,101 @@
+Thread-assisted mode of operation.
+
+Constraints
+-----------
+
+We need to keep using the existing ERR API as doing otherwise would
+complicate the existing applications and break our API compatibility promise.
+Even the ERR_STATE structure is public, although deprecated, and thus its
+structure and semantics cannot be changed.
+
+The error stack access is not under a lock (because it is thread-local).
+This complicates _moving errors between threads_.
+
+Error stack entries contain allocated data, copying entries between threads
+would mean duplicating it or losing it.
+
+Assumptions
+-----------
+
+This document assumes the error state of the QUIC connection (or stream for
+stream level errors) is handled separately.
+
+We can assume the internal assistance thread is well-behaving in regards
+to the error stack.
+
+We assume there are two types of errors that can be raised in the QUIC
+library calls and in the subordinate libcrypto (and provider) calls. First
+type is an intermittent error that does not really affect the state of the
+QUIC connection - for example EAGAIN returned on a syscall, or unavailability
+of some algorithm where there are other algorithms to try. Second type
+is a permanent error that affects the error state of the QUIC connection.
+Operations on QUIC streams (SSL_write(), SSL_read()) can also trigger errors,
+depending on their effect they are either permanent if they cause the
+QUIC connection to enter an error state, or if they just affect the stream
+they are left on the error stack of the thread that called SSL_write()
+or SSL_read() on the stream.
+
+Design
+------
+
+Return value of SSL_get_error() on QUIC connections or streams does not
+depend on the error stack contents.
+
+Intermittent errors are handled within the library and cleared from the
+error stack before returning to the user.
+
+Permanent errors happenning within the assist thread need to be transferred
+to the regular user thread. To simplify the implementation they are required
+to appear on the error stack reachable by the application ERR_ calls only
+when the SSL_get_error() is called from the application.
+
+Implementation
+--------------
+
+There is an error stack in QUIC_CHANNEL which serves as temporary storage
+for errors happening in the internal assistance thread. When a permanent error
+is detected in the assistance thread the error stack entries are copied
+to this error stack in QUIC_CHANNEL.
+
+When SSL_get_error() is called and there is a permanent error condition
+detected the errors from the QUIC_CHANNEL error stack are moved to the
+error stack of the thread that is calling SSL_get_error() on the SSL
+object representing the QUIC_CONNECTION.
+
+SSL_tick() return value
+-----------------------
+
+If a permanent error is detected during an SSL_tick() call, SSL_tick()
+should return an error to make an application be aware of the error
+condition. Otherwise an application keeping an idle connection with
```

Comment:
There is a problem here. If a connection fails, for example due to a protocol violation or for that matter a normal connection closure by an application, the connection enters the Terminating state. However, this doesn't mean you can't still tick the connection. There are shutdown phases in a QUIC connection which take time. For example, if a CONNECTION_CLOSE frame we sent is lost in the network, we need to send one again when the peer keeps sending normal traffic to us. etc. So the QUIC reactor actually needs to be able to keep functioning even in the Terminating state. Even if the termination was initiated as a result of a ERR on our end which has been stashed in the QUIC_CHANNEL.

Once we reach the Terminat*ed* state, I guess SSL_tick could return 0.

In general SSL_tick return values is not a good way to signal connection failure to an application, the failure of SSL_write/etc. (which we already do) is a better way. But open to suggestions on a solution.

Reply:
Could we make the SSL_tick() return value tri-state? I.e., 1 - everything is OK, 0 - connection is terminated, no need to call again. -1 - connection is terminating due to permanent error.

Reply:
That feels a bit awkward to me... I suspect @t-j-h wouldn't be a massive fan either. It might be more ergonomic to provide a function like `SSL_is_terminating`, so `SSL_tick` continues to return 1 while terminating but the application can bail early if it wishes. I might even go so far as to argue `SSL_tick` should still return 1 in the terminat*ed* state (a no-op state isn't a failure) and provide `SSL_is_terminated`. Then `SSL_tick` (or `SSL_handle_events` as it is about to be named) would only return 0 if some severe internal issue occurs (allocation error, etc.)

Thoughts?